File: tests/high_priority_regressions.rs
use minisearch_wasm::{MiniSearch, MiniSearchOptions, PartialSearchOptions, SearchOptions};
use serde_json::{json, Value};

fn engine() -> MiniSearch {
    MiniSearch::new(
        serde_json::from_value::<MiniSearchOptions>(json!({
            "fields": ["text", "optional"], "storeFields": ["payload"], "autoVacuum": false,
            "searchOptions": {"prefix": true, "fuzzy": 0.2}
        }))
        .unwrap(),
    )
}
fn snapshot(search: &MiniSearch) -> Value {
    serde_json::to_value(search).unwrap()
}
fn reloaded(search: &MiniSearch) -> Vec<MiniSearch> {
    vec![
        MiniSearch::from_bytes(&search.to_bytes().unwrap()).unwrap(),
        MiniSearch::from_json(&serde_json::to_string(search).unwrap()).unwrap(),
    ]
}

#[test]
fn field_presence_and_boundary_empty_tokens_survive_reload_and_discard() {
    let mut search = engine();
    search
        .add_all(vec![
            json!({"id":1,"text":"apple"}),
            json!({"id":2,"text":" apple "}),
            json!({"id":3,"text":null}),
            json!({"id":4}),
            json!({"id":5,"text":""}),
            json!({"id":6,"text":" ... "}),
        ])
        .unwrap();
    assert_eq!(
        snapshot(&search)["field_length"],
        json!([1, 0, 2, 0, 0, 0, 0, 0, 1, 0, 1, 0])
    );
    assert_eq!(
        snapshot(&search)["field_present"],
        json!([true, false, true, false, false, false, false, false, true, false, true, false])
    );
    let averages = snapshot(&search)["average_field_length"].clone();
    for mut index in reloaded(&search) {
        index.discard(&json!(4)).unwrap();
        index.discard(&json!(3)).unwrap();
        assert_eq!(snapshot(&index)["average_field_length"], averages);
        assert_eq!(index.search("apple", SearchOptions::default()).len(), 2);
        index.vacuum();
        assert_eq!(snapshot(&index)["average_field_length"], averages);
    }
}

#[test]
fn long_field_lengths_are_exact_when_discarded_after_reload() {
    let text = (0..70_000)
        .map(|i| format!("t{i}"))
        .collect::<Vec<_>>()
        .join(" ");
    let mut search = engine();
    search
        .add_all(vec![
            json!({"id":1,"text":text}),
            json!({"id":2,"text":"t0"}),
        ])
        .unwrap();
    assert_eq!(snapshot(&search)["field_length"][0], 70_000);
    for mut index in reloaded(&search) {
        index.discard(&json!(1)).unwrap();
        assert_eq!(snapshot(&index)["average_field_length"][0], 1.0);
        assert_eq!(index.search("t0", SearchOptions::default())[0].id, 2);
    }
}

#[test]
fn ordered_prefix_fuzzy_results_and_suggestions_survive_mutations_and_snapshots() {
    for words in [
        "apply application apple",
        "car cat can",
        "dog data day",
        "éclair écran école",
    ] {
        let mut search = engine();
        search
            .add_all(vec![
                json!({"id":1,"text":words}),
                json!({"id":2,"text":"apparatus carrot date"}),
                json!({"id":3,"text":"app car dog"}),
            ])
            .unwrap();
        for phase in 0..4 {
            match phase {
                1 => search
                    .remove(&json!({"id":2,"text":"apparatus carrot date"}))
                    .unwrap(),
                2 => {
                    search.discard(&json!(3)).unwrap();
                    search.add(json!({"id":4,"text":"cat app dog"})).unwrap();
                }
                3 => search.vacuum(),
                _ => {}
            }
            for restored in reloaded(&search) {
                for query in ["a", "c", "d", "é", "appl", "dag", "car app"] {
                    assert_eq!(
                        restored.search_joined_default(query, false),
                        search.search_joined_default(query, false),
                        "phase={phase}, query={query}"
                    );
                    assert_eq!(
                        restored.auto_suggest(query, None),
                        search.auto_suggest(query, None)
                    );
                }
                assert_eq!(restored.to_bytes().unwrap(), search.to_bytes().unwrap());
            }
        }
    }
}

#[test]
fn compact_ids_preserve_json_values_and_generation_changes() {
    let mut search = engine();
    let ids = vec![
        json!("a\nb"),
        json!(1),
        json!("1"),
        json!(""),
        json!(false),
        json!(["x", 1]),
        json!({"a":"b\nc"}),
    ];
    for id in &ids {
        search.add(json!({"id":id,"text":"apple"})).unwrap();
    }
    for mut index in reloaded(&search) {
        let joined: Vec<Value> =
            serde_json::from_str(&index.search_joined_default("apple", false).ids).unwrap();
        assert_eq!(joined, ids);
        let mut version = index.id_table_version();
        let raw = index.search_raw("apple", &PartialSearchOptions::default());
        assert_eq!(raw.id_table_version, version);
        let table: Vec<Value> = serde_json::from_str(&index.doc_id_table()).unwrap();
        assert_eq!(
            raw.doc_ids
                .iter()
                .map(|id| table[*id as usize].clone())
                .collect::<Vec<_>>(),
            ids
        );
        index.discard(&json!(1)).unwrap();
        assert_ne!(index.id_table_version(), version);
        assert_eq!(
            serde_json::from_str::<Value>(&index.doc_id_table()).unwrap()[1],
            Value::Null
        );
        version = index.id_table_version();
        index.add(json!({"id":1,"text":"apple"})).unwrap();
        assert_ne!(index.id_table_version(), version);
        version = index.id_table_version();
        index.remove_all_documents();
        assert_ne!(index.id_table_version(), version);
        assert_eq!(index.doc_id_table(), "[]");
        assert_eq!(index.search_joined_default("apple", false).ids, "[]");
    }
}

fn valid_snapshot() -> Value {
    let mut search = engine();
    search
        .add(json!({"id":"first", "text":"apple", "payload":{"tag": "ok"}}))
        .unwrap();
    snapshot(&search)
}

#[test]
fn json_rejects_inconsistent_state_including_direct_serde_deserialization() {
    let edits = [
        ("/snapshot_version", json!(3)),
        ("/document_count", json!(900)),
        ("/next_id", json!(0)),
        ("/next_id", json!(u32::MAX)),
        ("/dirt_count", json!(2)),
        ("/field_ids/text", json!(100)),
        ("/field_length", json!([])),
        ("/field_present", json!([])),
        ("/average_field_length", json!([])),
        ("/average_field_length/0", json!(-1)),
        ("/field_present/0", json!(false)),
        ("/id_to_short_id/s:first", json!(7)),
        ("/document_ids/0", Value::Null),
        ("/index/root/children/0/0", json!("")),
        ("/index/root/children/0/1/leaf_pos", json!(99)),
        ("/index/root/children/0/1/leaf/0/0", json!(0)),
    ];
    for (path, value) in edits {
        let mut corrupt = valid_snapshot();
        *corrupt
            .pointer_mut(path)
            .unwrap_or_else(|| panic!("missing test target {path}")) = value;
        assert!(
            MiniSearch::from_json(&corrupt.to_string()).is_err(),
            "accepted {path}"
        );
        assert!(
            serde_json::from_value::<MiniSearch>(corrupt).is_err(),
            "serde accepted {path}"
        );
    }
    let mut unknown = valid_snapshot();
    unknown["stored_fields"]["99"] = json!({"payload":1});
    assert!(MiniSearch::from_json(&unknown.to_string()).is_err());
    let mut unversioned = valid_snapshot();
    unversioned
        .as_object_mut()
        .unwrap()
        .remove("snapshot_version");
    assert!(MiniSearch::from_json(&unversioned.to_string()).is_err());
}

#[test]
fn stale_postings_require_dirt_but_valid_dirty_indexes_load() {
    let mut search = engine();
    search
        .add_all(vec![
            json!({"id":1,"text":"apple"}),
            json!({"id":2,"text":"pear"}),
        ])
        .unwrap();
    search.discard(&json!(1)).unwrap();
    assert_eq!(reloaded(&search).len(), 2);
    let mut corrupt = snapshot(&search);
    corrupt["dirt_count"] = json!(0);
    assert!(MiniSearch::from_json(&corrupt.to_string()).is_err());
}

#[test]
fn truncated_mutated_and_oversized_binary_inputs_never_panic() {
    let search = MiniSearch::from_json(&valid_snapshot().to_string()).unwrap();
    let bytes = search.to_bytes().unwrap();
    for length in 0..bytes.len() {
        assert!(
            MiniSearch::from_bytes(&bytes[..length]).is_err(),
            "prefix length {length}"
        );
    }
    let mut trailing = bytes.clone();
    trailing.push(0);
    assert!(MiniSearch::from_bytes(&trailing)
        .unwrap_err()
        .contains("trailing"));
    // Every single byte is corrupted with three masks; accepted mutations must
    // still support search and another validated round trip without trapping.
    for offset in 0..bytes.len() {
        for mask in [1, 0x80, 0xff] {
            let mut corrupt = bytes.clone();
            corrupt[offset] ^= mask;
            let attempt = std::panic::catch_unwind(|| {
                if let Ok(loaded) = MiniSearch::from_bytes(&corrupt) {
                    loaded.search_joined_default("apple", false);
                    MiniSearch::from_bytes(&loaded.to_bytes().unwrap()).unwrap();
                }
            });
            assert!(attempt.is_ok(), "panic at byte {offset}, mask {mask}");
        }
    }
    for corrupt in [
        vec![0xff; 10],
        vec![0x84, 0],
        vec![4, 0xff, 0xff, 0xff, 0xff, 0x7f],
    ] {
        assert!(MiniSearch::from_bytes(&corrupt).is_err());
    }
}

File: differential/high_priority_regressions.mjs
// End-to-end regressions for the four high-priority review findings.
// Run after npm run build. The pinned JS engine is the ranking oracle.
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import MiniSearch from 'minisearch'
import init, { MiniSearchWasm } from '../pkg/minisearch_wasm_core.js'

await init({ module_or_path: readFileSync(new URL('../pkg/minisearch_wasm_bg.wasm', import.meta.url)) })
let assertions = 0
const equal = (actual, expected, label) => { assert.deepEqual(actual, expected, label); assertions++ }
const options = { fields: ['text', 'optional'], storeFields: ['payload'], autoVacuum: false }
const key = id => JSON.stringify(id)
const decodeJoined = r => {
  const ids = JSON.parse(r.ids)
  equal(ids.length, r.count, 'joined row count')
  const terms = r.count ? r.terms.split('\n') : []
  return ids.map((id, i) => ({ id, score: r.scores[i], terms: terms[i] ? terms[i].split(' ') : [] }))
}
const decodeRaw = (engine, r) => {
  equal(r.idTableVersion, engine.idTableVersion, 'raw generation')
  const table = JSON.parse(engine.docIdTable())
  const terms = r.termTable ? r.termTable.split('\n') : []
  return Array.from(r.docIds, (id, i) => ({ id: table[id], score: r.scores[i],
    terms: Array.from(r.termIds.slice(r.termOffsets[i], r.termOffsets[i + 1]), term => terms[term]) }))
}
function parity(js, wasm, label, queries = ['apple', 'app', 'applf', 'pear', 'apple pear', '']) {
  equal(wasm.documentCount, js.documentCount, label + ' document count')
  for (const query of queries) for (const opts of [{}, { prefix: true }, { fuzzy: 0.2 }, { prefix: true, fuzzy: 0.2, combineWith: 'AND' }]) {
    js.search(query, opts) // JS's documented dirty-index cleanup fixpoint.
    const expected = js.search(query, opts)
    const byId = new Map(expected.map(row => [key(row.id), row]))
    for (const rows of [wasm.search(query, opts), decodeJoined(wasm.searchJoinedOpts(query, opts)), decodeRaw(wasm, wasm.searchRaw(query, opts))]) {
      equal(rows.length, expected.length, label + ' result count')
      for (const row of rows) {
        const other = byId.get(key(row.id))
        assert.ok(other, label + ' missing ID ' + key(row.id))
        const delta = Math.abs(row.score - other.score) / Math.max(Math.abs(row.score), Math.abs(other.score), 1e-12)
        assert.ok(Number.isFinite(row.score) && delta <= 1e-12, `${label}: ${query} score ${row.score} vs ${other.score}`)
        equal([...row.terms].sort(), [...other.terms].sort(), label + ' matched terms')
      }
    }
  }
}
function roundTrips(engine, check) {
  for (const copy of [MiniSearchWasm.loadBytes(engine.toBytes()), MiniSearchWasm.loadNativeJSON(engine.toNativeJSONString())]) {
    try { check(copy) } finally { copy.free() }
  }
}

// Sparse fields, empty boundary tokens, null fields, and removals/discards.
const documents = [
  { id: 1, text: 'apple', optional: 'pear', payload: { tag: 'first' } },
  { id: 2, text: ' apple ' }, { id: 3, text: null }, { id: 4 },
  { id: 5, text: '' }, { id: 6, text: ' ... ' },
  { id: 7, text: 'apple,,pear!', optional: null }, { id: 8, text: 'pear' }
]
let seed = 0x92119
const random = n => { seed = (Math.imul(seed, 1664525) + 1013904223) >>> 0; return seed % n }
const fields = [undefined, null, '', ' ', 'apple', ' apple ', '.pear apple!', 'apple...pear', 'pear', []]
for (let i = 0; i < 100; i++) {
  const doc = { id: i + 20 }
  for (const name of options.fields) { const value = fields[random(fields.length)]; if (value !== undefined) doc[name] = value }
  documents.push(doc)
}
const js = new MiniSearch(options), wasm = new MiniSearchWasm(options)
js.addAll(documents); wasm.addAll(documents)
for (const phase of ['fresh', 'removed', 'discarded', 'vacuumed']) {
  if (phase === 'removed') {
    for (const doc of documents.filter((_, i) => i % 5 === 0)) { js.remove(doc); wasm.remove(doc) }
  } else if (phase === 'discarded') {
    for (const doc of documents.filter((_, i) => i % 5 === 1)) { js.discard(doc.id); wasm.discard(doc.id) }
  } else if (phase === 'vacuumed') { await js.vacuum(); await wasm.vacuum() }
  parity(js, wasm, phase)
  roundTrips(wasm, copy => parity(js, copy, phase + ' reloaded'))
}
wasm.free()
console.log('ok field accounting: seeded sparse fields, all result formats, maintenance and both snapshots')

// Above the former u16 limit; compare scores both before and after discards.
const longDocs = [65535, 65536, 70000].map((length, id) => ({ id,
  text: Array.from({ length }, (_, i) => 't' + i).join(' ') }))
longDocs.push({ id: 3, text: 't0' })
const longJS = new MiniSearch(options), longWasm = new MiniSearchWasm(options)
longJS.addAll(longDocs); longWasm.addAll(longDocs)
parity(longJS, longWasm, 'long fields', ['t0'])
roundTrips(longWasm, copy => {
  const reference = new MiniSearch(options); reference.addAll(longDocs)
  for (let id = 0; id < 3; id++) { reference.discard(id); copy.discard(id) }
  equal(copy.toNativeJSON().average_field_length[0], 1, 'average after discarding long fields')
  parity(reference, copy, 'long fields discarded after reload', ['t0'])
})
longWasm.free()
console.log('ok exact field lengths at 65535, 65536 and 70000 unique tokens')

// Preserve complete suggestions, term ordering and scores, including splits,
// leaf/child interleaving, Unicode edges, deletes and dirty postings.
for (const text of ['apply application apple', 'app apply application apple', 'car cat can', 'dog data day', 'éclair écran école']) {
  const engine = new MiniSearchWasm({ ...options, searchOptions: { prefix: true, fuzzy: 0.2 } })
  engine.addAll([{ id: 1, text }, { id: 2, text: 'app cart dog' }, { id: 3, text: 'apparatus carrot day' }])
  for (let phase = 0; phase < 4; phase++) {
    if (phase === 1) engine.remove({ id: 3, text: 'apparatus carrot day' })
    if (phase === 2) engine.discard(2)
    if (phase === 3) await engine.vacuum()
    roundTrips(engine, copy => {
      for (const query of ['a', 'app', 'c', 'd', 'é', 'applf', 'dag']) {
        equal(copy.searchJoined(query, false), engine.searchJoined(query, false), 'ordered joined round trip')
        equal(copy.search(query), engine.search(query), 'full round trip')
        equal(copy.autoSuggest(query), engine.autoSuggest(query), 'suggestion round trip')
      }
    })
  }
  engine.free()
}
console.log('ok radix traversal and full suggestion preservation across snapshots')

const ids = ['a\nb', 1, '1', '', true, ['a', 1], { id: 'nested\nvalue' }, '"quoted"\\slash', '😀']
const idEngine = new MiniSearchWasm(options)
idEngine.addAll(ids.map(id => ({ id, text: 'apple' })))
const verifyIds = engine => {
  const expected = engine.search('apple').map(row => row.id)
  equal(decodeJoined(engine.searchJoined('apple', false)).map(row => row.id), expected, 'joined typed IDs')
  equal(decodeRaw(engine, engine.searchRaw('apple')).map(row => row.id), expected, 'raw typed IDs')
}
verifyIds(idEngine); roundTrips(idEngine, verifyIds)
let generation = idEngine.idTableVersion
idEngine.discard(1)
assert.notEqual(idEngine.idTableVersion, generation)
equal(JSON.parse(idEngine.docIdTable())[1], null, 'discard leaves null table slot')
generation = idEngine.idTableVersion
idEngine.add({ id: 1, text: 'apple' })
assert.notEqual(idEngine.idTableVersion, generation)
verifyIds(idEngine); roundTrips(idEngine, verifyIds)
generation = idEngine.idTableVersion
idEngine.remove({ id: true, text: 'apple' })
assert.notEqual(idEngine.idTableVersion, generation)
generation = idEngine.idTableVersion
idEngine.removeAll()
assert.notEqual(idEngine.idTableVersion, generation)
equal(JSON.parse(idEngine.docIdTable()), [], 'reset empty table')
equal(JSON.parse(idEngine.searchJoined('apple', false).ids), [], 'empty joined IDs')
assert.throws(() => idEngine.add({ id: null, text: 'apple' }))
idEngine.free()
console.log('ok lossless compact IDs, null rejection and mutation generations')

// Small deliberately malformed v4 inputs must throw descriptive errors, never
// a WebAssembly.RuntimeError. Huge advertised lengths must fail before reserve.
const v = n => { const out = []; do { const byte = n % 128; n = Math.floor(n / 128); out.push(byte | (n ? 128 : 0)) } while (n); return out }
const str = text => { const bytes = [...new TextEncoder().encode(text)]; return [...v(bytes.length), ...bytes] }
const float = n => { const bytes = new Uint8Array(8); new DataView(bytes.buffer).setFloat64(0, n, true); return [...bytes] }
const header = (count = 0, next = count, dirt = 0, average = 0) => [4, ...str('{"fields":["text"]}'), ...v(count), ...v(next), ...v(dirt), ...float(average)]
const document = [0, 3, 1, 2, 0] // short ID delta, tagged unsigned ID 1, present length, stored count
const leaf = [0, 1, 0, 1, 0, 1, 0] // leaf slot; field; one posting; no children
const valid = [...header(1, 1, 0, 1), ...document, 0, 0, 1, ...str('apple'), ...leaf]
const probe = MiniSearchWasm.loadBytes(new Uint8Array(valid)); probe.free()
let deep = leaf
for (let i = 0; i < 130; i++) deep = [0, 0, 1, ...str('a'), ...deep]
const malformed = [
  [3], [0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x7f], [0x84, 0],
  [4, ...v(0xffffffff)], [...header(0, 0xffffffff), 0, 0, 0],
  [...header(0, 0, 0, NaN), 0, 0, 0], [...valid, 0],
  [...header(1, 1, 0, 1), ...document, ...deep],
  [...header(), 0, 0, 1, 1, 0xff, 0, 0, 0],
  [...header(1, 1, 0, 1), ...document, 0, 0, 2, ...str('a'), ...leaf, ...str('ab'), ...leaf],
  [...header(1, 1, 0, 1), ...document, 0, 0, 1, ...str('a'), 99, ...leaf.slice(1)]
]
for (const bytes of malformed) {
  assert.throws(() => MiniSearchWasm.loadBytes(new Uint8Array(bytes)), error => error instanceof Error && !(error instanceof WebAssembly.RuntimeError))
  assertions++
}
for (let length = 0; length < valid.length; length++) {
  assert.throws(() => MiniSearchWasm.loadBytes(new Uint8Array(valid.slice(0, length))), error => error instanceof Error)
}
const validEngine = MiniSearchWasm.loadBytes(new Uint8Array(valid))
for (const [field, value] of [['document_count', 900], ['next_id', 0], ['field_present', [false]], ['average_field_length', [-1]], ['snapshot_version', 3]]) {
  const state = validEngine.toNativeJSON(); state[field] = value
  assert.throws(() => MiniSearchWasm.loadNativeJSON(JSON.stringify(state)), error => error instanceof Error)
}
for (let i = 0; i < valid.length; i++) {
  const bytes = new Uint8Array(valid); bytes[i] ^= 0x80
  let copy
  try { copy = MiniSearchWasm.loadBytes(bytes) } catch (error) { assert.ok(error instanceof Error); continue }
  copy.searchJoined('apple', false); copy.free()
}
equal(validEngine.search('apple')[0].id, 1, 'valid index remains usable after invalid loads')
validEngine.free()
console.log('ok malformed snapshots, overflow, recursion, truncation and mutation probes through WASM')
console.log(`HIGH PRIORITY REGRESSIONS: ALL PASS (${assertions} explicit equality/rejection checks)`)
